tlv/record.go

```diff
@@ -0,0 +1,147 @@
+package tlv
+
+import (
+	"io"
+
+	"github.com/btcsuite/btcd/btcec"
+)
+
+// Type is an 64-bit identifier for a TLV Record.
+type Type uint64
+
+// Encoder is a signature for methods that can encode TLV values. An error
+// should be returned if the Encoder cannot support the underlying type of val.
+// The provided scratch buffer must be non-nil.
+type Encoder = func(w io.Writer, val interface{}, buf *[8]byte) error
+
+// Decoder is a signature for methods that can decode TLV values. An error
+// should be returned if the Decoder cannot support the underlying type of val.
+// The provided scratch buffer must be non-nil.
+type Decoder = func(r io.Reader, val interface{}, buf *[8]byte, l uint64) error
+
+// SizeFunc is a function that can compute the length of a given field. Since
+// the size of the underlying field can change, this allows the size of the
+// field to be evaluated at the time of encoding.
+type SizeFunc func() uint16
+
+// SizeVarBytes is a SizeFunc that can compute the length of a byte slice.
+func SizeVarBytes(e *[]byte) SizeFunc {
+	return func() uint16 {
+		return uint16(len(*e))
+	}
+}
+
+// Record holds the required information to encode or decode a TLV record.
+type Record struct {
+	typ        Type
+	staticSize uint16
+	value      interface{}
+	sizeFunc   SizeFunc
+	encoder    Encoder
+	decoder    Decoder
+}
+
+// Size returns the size of the Record's value. If no static size is known, the
+// dynamic size will be evaluated.
+func (f *Record) Size() uint64 {
+	if f.sizeFunc == nil {
+		return uint64(f.staticSize)
+	}
+
+	return uint64(f.sizeFunc())
+}
+
+// MakePrimitiveRecord creates a record for common types.
+func MakePrimitiveRecord(typ Type, val interface{}) Record {
+	var (
+		staticSize uint16
+		sizeFunc   SizeFunc
+		encoder    Encoder
+		decoder    Decoder
+	)
+	switch e := val.(type) {
+	case *uint8, *int8:
+		staticSize = 1
+		encoder = EUint8
```

Comment:
I think the commits were re-ordered? This type is used before its defined in the series.

Reply:
i think it might just be a gh thing? the commits look in order locally